File: nmLabReport/lab34.c
#include <stdio.h>
#include <math.h>

#define N 10
#define MAX_ITER 1000
#define TOL 1e-5

void initialize(double grid[N][N]) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (i == 0 || i == N-1 || j == 0 || j == N-1) {
                grid[i][j] = 100.0; // Boundary conditions
            } else {
                grid[i][j] = 0.0; // Initial guess
            }
        }
    }
}

void print_grid(double grid[N][N]) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("%6.2f ", grid[i][j]);
        }
        printf("\n");
    }
}

void solve_laplace(double grid[N][N]) {
    double new_grid[N][N];
    int iter = 0;
    double diff;

    do {
        diff = 0.0;
        for (int i = 1; i < N-1; i++) {
            for (int j = 1; j < N-1; j++) {
                new_grid[i][j] = 0.25 * (grid[i+1][j] + grid[i-1][j] + grid[i][j+1] + grid[i][j-1]);
                diff = fmax(diff, fabs(new_grid[i][j] - grid[i][j]));
            }
        }

        for (int i = 1; i < N-1; i++) {
            for (int j = 1; j < N-1; j++) {
                grid[i][j] = new_grid[i][j];
            }
        }

        iter++;
    } while (diff > TOL && iter < MAX_ITER);
}

int main() {
    double grid[N][N];
    initialize(grid);
    solve_laplace(grid);
    print_grid(grid);
    return 0;
}

File: nmLabReport/lab34.py
import numpy as np

N = 10
MAX_ITER = 1000
TOL = 1e-5

def initialize(grid):
    for i in range(N):
        for j in range(N):
            if i == 0 or i == N-1 or j == 0 or j == N-1:
                grid[i][j] = 100.0  # Boundary conditions
            else:
                grid[i][j] = 0.0  # Initial guess

def print_grid(grid):
    for row in grid:
        print(" ".join(f"{val:.2f}" for val in row))

def solve_laplace(grid):
    new_grid = np.copy(grid)
    iter = 0
    diff = 0.0

    while iter < MAX_ITER:
        diff = 0.0
        for i in range(1, N-1):
            for j in range(1, N-1):
                new_grid[i][j] = 0.25 * (grid[i+1][j] + grid[i-1][j] + grid[i][j+1] + grid[i][j-1])
                diff = max(diff, abs(new_grid[i][j] - grid[i][j]))

        grid = np.copy(new_grid)
        iter += 1

        if diff < TOL:
            break

def main():
    grid = np.zeros((N, N))
    initialize(grid)
    solve_laplace(grid)
    print_grid(grid)

if __name__ == "__main__":
    main()